tests/federation_room_join_partial_state_test.go

```diff
@@ -142,75 +146,116 @@ func TestPartialStateJoin(t *testing.T) {
-		stateReq := gomatrixserverlib.NewFederationRequest("GET", "hs1",
-			fmt.Sprintf("/_matrix/federation/v1/state_ids/%s?event_id=%s",
-				url.PathEscape(psjResult.ServerRoom.RoomID),
-				url.QueryEscape(event.EventID()),
-			),
-		)
-		var respStateIDs gomatrixserverlib.RespStateIDs
-		if err := psjResult.Server.SendFederationRequest(deployment, stateReq, &respStateIDs); err != nil {
-			t.Errorf("/state_ids request returned non-200: %s", err)
-			return
-		}
-		var gotState, expectedState []interface{}
-		for _, ev := range respStateIDs.StateEventIDs {
-			gotState = append(gotState, ev)
-		}
-		for _, ev := range psjResult.ServerRoom.AllCurrentState() {
-			expectedState = append(expectedState, ev.EventID())
-		}
-		must.CheckOffAll(t, gotState, expectedState)
+		// the HS will make a /get_missing_events request for the missing prev events of event B
+		handleGetMissingEventsRequests(t, psjResult.Server, psjResult.ServerRoom, []*gomatrixserverlib.Event{eventA})
+
+		// send event B to hs1
+		testReceiveEventDuringPartialStateJoin(t, deployment, alice, psjResult, eventB)
+	})
+
+	// we should be able to receive events with partially missing prevs over federation during the resync
+	t.Run("CanReceiveEventsWithHalfMissingPrevsDuringPartialStateJoin", func(t *testing.T) {
+		deployment := Deploy(t, b.BlueprintAlice)
+		defer deployment.Destroy(t)
+		alice := deployment.Client(t, "hs1", "@alice:hs1")
+
+		psjResult := beginPartialStateJoin(t, deployment, alice)
+		defer psjResult.Destroy()
+
+		// we construct the following event graph:
+		//         +---------+
+		//         v          \
+		// ... <-- M <-- A <-- B
+		//
+		// M is @alice:hs1's join event.
+		// A and B are regular m.room.messsage events sent by @derek from Complement.
+		//
+		// initially, hs1 only knows about event M.
+		// we send only event B to hs1.
+		eventM := psjResult.ServerRoom.CurrentState("m.room.member", alice.UserID)
+		eventA := psjResult.CreateMessageEvent(t, "derek", []string{eventM.EventID()})
+		eventB := psjResult.CreateMessageEvent(t, "derek", []string{eventA.EventID(), eventM.EventID()})
+		t.Logf("%s's m.room.member event is %s", *eventM.StateKey(), eventM.EventID())
+		t.Logf("Derek sent event A with ID %s", eventA.EventID())
+		t.Logf("Derek sent event B with ID %s", eventB.EventID())
+
+		// the HS will make an /event_auth request for event A
+		federation.HandleEventAuthRequests()(psjResult.Server)
+
+		// the HS will make a /get_missing_events request for the missing prev event of event B
+		handleGetMissingEventsRequests(t, psjResult.Server, psjResult.ServerRoom, []*gomatrixserverlib.Event{eventA})
+
+		// send event B to hs1
+		testReceiveEventDuringPartialStateJoin(t, deployment, alice, psjResult, eventB)
+	})
+
+	// we should be able to receive events with missing prevs over federation during the resync
+	t.Run("CanReceiveEventsWithMissingPrevWithHalfMissingPrevsDuringPartialStateJoin", func(t *testing.T) {
```

Comment:
Suggestions for better names are welcome.

Reply:
maybe "can receive events with some missing grandparents" or something?